feat: text labels

flekschas opened the pull request:
This PR adds support for text labels via a new function: `scatter.label()`

https://github.com/user-attachments/assets/5a145970-436d-4680-9e30-65ab444e384e

## Description

### What was changed in this PR?

1. Added new `scatter.label()` function for displaying text labels
2. Added `LabelPlacement` class that allows to statically-resolved label placement and preserve results
3. Added new color helper methods: brighten, darken, saturate, desaturate
4. Added 2D color map support via [pycolormap-2d](https://pypi.org/project/pycolormap-2d/)

### Example

For in-depth examples, see https://github.com/flekschas/jupyter-scatter-tutorial/blob/main/notebooks/6-Labels.ipynb. In the following I provide a short example.

```py
import jscatter
import pandas as pd

df_cities = pd.read_parquet(https://storage.googleapis.com/flekschas/jupyter-scatter-tutorial/cities.pq)

scatter_cities = jscatter.Scatter(
    data=df_cities,
    x='Mercator X',
    y='Mercator Y',
    color_by='Continent',
    size=2,
    axes=False,
    legend=True,
    height=720,
    # The column by which we want to label points
    label_by='Continent',
)

scatter_cities.show()
```

### Why is this PR necessary?

Fixes #73 

## Checklist

- [x] Provided a concise title as a [semantic commit message](https://www.conventionalcommits.org) (e.g. "fix: correctly handle undefined properties")
- [x] `CHANGELOG.md` updated
- [x] Tests added or updated
- [x] Documentation in `README.md` added or updated
- [x] Example(s) added or updated
- [x] Screenshot, gif, or video attached for visual changes

manzt commented on jscatter/label_placement/center_of_mass.py:
```diff
@@ -0,0 +1,105 @@
+import numpy as np
+import numpy.typing as npt
+from numba import njit
```
Does this mean numba is now a dependency of jscatter? I could imagine that would make it challenging to use on newer versions of Python.

flekschas replied:
Yeah, numbs is a dependency now, which already led to issues with Python v3.13 for which I had to drop support because Numba doesn't yet support it. I might do a performance test against the plain Python version to see if the jit compiled methods actually boost the performance noticeably.

manzt commented on pyproject.toml:
```diff
@@ -32,16 +31,20 @@ keywords = [
   "anywidget>=0.9.0",
+  "fast_hdbscan",
+  "geoindex-rs",
   "ipywidgets>=7.6,<9",
   "ipython",
-  "matplotlib",
+  "matplotlib>=3.9",
+  "numba",
   "numpy",
   "pandas",
   "pooch>=1.3.0",
   "pyarrow",
+  "pycolormap-2d",
   "scipy",
   "seaborn",
-  "traittypes>=0.2.1",
+  "tqdm",
```
Not a huge deal, but I'm wondering if there is a way to feature flag these additional dependencies with an `project.optional-dependencies`. Something like:

```sh
pip install "jupyter-scatter[all]"
```

or 

```sh
pip install "jupyter-scatter[labels]"
```

and then have a runtime error if someone enters a code path with a dependency that is required for a specific feature.

flekschas replied:
I'm all for it. I think almost all labeling features work without any new additions other than the geoindex which is lightweight.

Do you know how we would raise errors appropriately for features that would require optional dependencies?

manzt reviewed (commented):
This is awesome. I tried it out and works really well, super excited to see this feature come to life. I like the high-level API _a lot_. 

My main comment is that this specific feature seems to add a few dependencies and constraints on what environments jupyter-scatter can be run in. Perhaps for a follow up PR, one idea would be to move some of the dependencies under an "extras" flag, such that users can opt-in to the batteries included version of the package, and default to something a little more minimal.

flekschas commented:
@manzt What do you think about the way I've split the dependencies? There are now 4 option:
1. default
2. `all`: includes all extras below
3. `annotation-extras`: includes Seaborn for the contour annotation
4. `label-extras`: includes HDBSCAN for `largest_cluster` positioning and tqdm for showing label computation progress

I've also removed Numba is it didn't actually improve the performance. As part of removing Numba as a dependency I've also replaced `fast-hdbscan` with the regular `hdbscan` implementation as `fast-hdbscan` relies on Numba and is actually slow to import.
